### DISCUSSION TITLE
Terraform lockfile zh hashes

### DISCUSSION THREAD

**Discussion by viceice** [renovatebot/renovate#11821] [Suggest an Idea]
Currently Renovate updates only the `h1` hashes. So when i do an `terraform init` i now get a modified lockfile
![image](https://user-images.githubusercontent.com/1798109/134000053-b3cf1917-74b1-4336-aa52-d57ecd1ddf57.png)

@secustor  Is it possible to additionally generate the `zh` hashes?

**Comment by viceice** [comment]
Or is there maybe an option to suppress the old `zh` hashes 🤔

**Comment by secustor** [comment]
Yes it should be possible to add them as these are exposed using their API https://releases.hashicorp.com/terraform-provider-kubernetes/2.5.0/terraform-provider-kubernetes_2.5.0_SHA256SUMS, but they are not needed for the functionality.     

You should be able to skip this using `terraform init -lockfile=readonly`

**Reply by viceice** [reply]
trying ` -lockfile=readonly`

**Reply by viceice** [reply]
ok `-lockfile=readonly` seems to work, would like to have it as default 😅 

are the `zh` hases outdated? Will they be removed in future? or does it make sense to support the too?

**Reply by secustor** [reply]
`zh` hashes are only for the download. Terraform generates a `h1` for your local arch and os and complains if there is none. So `h1` make it work. 

This suggests that they are facing it out https://www.terraform.io/docs/language/dependency-lock.html#zh-  but no timeline as far as I know

**Comment by ekristen** [comment]
First renovate is **AMAZING**, probably one of the best OSS tools in well over a decade and what you've all done is amazing, so thank you!

I would like to jump in on this discussion. Unfortunately I have to disagree with your point of view @secustor.


It is my opinion that you are essentially saying that if someone wants to use renovate for terraform that they have to use non-default options to make it work, I think this is a bad precedence to set with renovate. 

To put this another way, in the context of node.js, this would be the equivalent to saying renovate can keep dependencies up to date, but the file has to be called `package-renovate.json` and you have to call a non-standard `npm install --config=renovate-package.json` to use it otherwise it won't work. Yes I'm aware that's not a valid command, just using as an example to make a point.

The suggestion to use the `-lockfile=readonly` is a workaround to prevent issues with a lockfile generated by renovate that is essentially incomplete.

Another reason I would consider this a workaround is that even though `zh` is considered legacy, there is no way to natively disable it from being added by terraform.

Until terraform stops using and adding the `zh` hash, renovate have parity on it's own implementation, otherwise it looks like all the PRs done by renovate are inconsistent with user usage.

Thank you for your time.

**Reply by secustor** [reply]
This is a valid request, feel free to open a feature request. 

If you think about contributing this feature, we would happely give you some pointers.

**Reply by ekristen** [reply]
Already opened. https://github.com/renovatebot/renovate/issues/13692 -- I'm happy to contribute, I didn't have good luck the last time I tried to figure out how to setup renovate for development purposes, I can take a look again though and see what I can do.

**Reply by rarkins** [reply]
@secustor can you remind me of the reasoning of whether to write our own lock file updating vs call a `terraform` binary? Wondering if we should switch to the "usual" way or if the custom updating saves us time or downloads?

**Reply by viceice** [reply]
The `terraform` binary has no option to do this without full access to backend. So renovate would need and have full administrative access to all terraform managed resources and the terraform state.

**Reply by viceice** [reply]
Original issue
- #7895

**Reply by viceice** [reply]
and the cause:
- https://github.com/hashicorp/terraform/issues/27161

**Reply by secustor** [reply]
Also this setup allows us to cache the hashes, which the CLI does not. 
 - https://github.com/hashicorp/terraform/issues/27811
 - https://github.com/hashicorp/terraform/issues/28333

**Reply by pascal-hofmann** [reply]
A workaround to prevent terraform from adding the `zh` hashes is setting the env var `TF_CLI_ARGS_init="-lockfile=readonly`.

This way not every call to `terraform init` has to be changed.

**Reply by Daniel-I-Am** [reply]
Missing `zh` hash format is still an issue when renovating Terraform. And Hashicorp still shows no sign of actually removing them from the lockfile.

The issue that was ostensibly created for this has been deleted (https://github.com/renovatebot/renovate/issues/13692), is there still an issue currently tracking this feature request?

**Reply by rarkins** [reply]
We don't have any solution or plan to change anything so therefore this Discussion is the place to track.

Solutions are welcome, repeating the problem doesn't really help at all

**Reply by Daniel-I-Am** [reply]
All good! I will take a stab at the problem, see if it's something I can figure out :crossed_fingers:
